### PR TITLE
3.0 update Cache in-memory

### LINKED ISSUE #10218: Update Cache in-memory
12 K PV/month

Moved from #6808

From @user135711 at https://github.com/aspnet/Docs/issues/5495#issuecomment-388458686 ...

> On the in memory Caching sample, please add an explanation of how cache dependencies work. I've been trying to figure it out from the source code and debugger, and the Docs don't provide sufficient explanation: 

```csharp
    using (var entry = _cache.CreateEntry(CacheKeys.Parent))
    {
        // expire this entry if the dependant entry expires.
        entry.Value = DateTime.Now;
        entry.RegisterPostEvictionCallback(DependentEvictionCallback, this);

        _cache.Set(CacheKeys.Child,
            DateTime.Now,
            new CancellationChangeToken(cts.Token));
    }
```

> How does the child token above communicate to the parent token to evict and what is the purpose of disposing of the parent token? Is this a hack where the parent eviction only occurs because the already marked as expired parent gets evicted when the child cancellation triggers a remove expired entries search?

> Second suggestion add to sample app a compatible version of this dependent cache entries construction because it makes more sense: https://github.com/aspnet/Caching/issues/48

**Additional**

* The sample app is MVC, not Razor Pages.
* The topic is in need of a UE pass.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 13f5f8f9-18dd-9c12-9ed1-fcc6b44ba7fd
* Version Independent ID: 640b4f36-5886-7b69-1c95-cfc8574a02fd
* Content: [Cache in-memory in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-2.2)
* Content Source: [aspnetcore/performance/caching/memory.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/performance/caching/memory.md)
* Product: **aspnet-core**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### LINKED ISSUE #10479: Update Get Started tutorial for 3.0
Update the tutorial for ASP.NET Core 3.0.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: ca6139d4-022c-395a-0abf-539660550df1
* Version Independent ID: 3dd23f8a-300e-3f78-1dc5-8b759aabdccc
* Content: [Get started with ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/getting-started/?view=aspnetcore-2.1&tabs=windows#next-steps)
* Content Source: [aspnetcore/getting-started/index.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/getting-started/index.md)
* Product: **aspnet-core**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### LINKED ISSUE #10858: Do you have an example of the absolute expiration time.
The article said:“The following sample: Sets the absolute expiration time.”
But I only found sliding expiration examples.
Please add the example or , if I missed it, highlight it by comment.

Ref:  https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory.cacheextensions.set?view=aspnetcore-2.2#Microsoft_Extensions_Caching_Memory_CacheExtensions_Set__1_Microsoft_Extensions_Caching_Memory_IMemoryCache_System_Object___0_System_TimeSpan_

@Rick-Anderson 	 edit:
16K PV, 83% CSAT

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 13f5f8f9-18dd-9c12-9ed1-fcc6b44ba7fd
* Version Independent ID: 640b4f36-5886-7b69-1c95-cfc8574a02fd
* Content: [Cache in-memory in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-2.2)
* Content Source: [aspnetcore/performance/caching/memory.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/performance/caching/memory.md)
* Product: **aspnet-core**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### LINKED ISSUE #12634: Why RegisterPostEvictionCallback is not invoked, when cache is expited
It will be good to add clarification on this page, why RegisterPostEvictionCallback is not invoked, when cache is expited 
It is explainined on https://github.com/aspnet/Caching/issues/248#issuecomment-258571802 Expiration almost never happens on it's own in the background #248
, but it’s useful enough to be included in documentation, as well as workaround:

Rather than using SetAbsoluteExpiration(TimeSpan.FromHours(1)) you could instead use CancellationTokenSource.CancelAfter(TimeSpan.FromHours(1)) for the token you're already registering. When this token fires it will remove the entry immediately and fire the eviction callbacks.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 13f5f8f9-18dd-9c12-9ed1-fcc6b44ba7fd
* Version Independent ID: 640b4f36-5886-7b69-1c95-cfc8574a02fd
* Content: [Cache in-memory in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-2.2)
* Content Source: [aspnetcore/performance/caching/memory.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/performance/caching/memory.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-performance**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes  #10479 Fixes #10218 Fixes #10858 Fixes #12634

[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812)

If there have been no changes in 3.0 to memory caching, then the review will be short.

Requires review: Does the  [System.Runtime.Caching/MemoryCache]( https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#systemruntimecachingmemorycache) section apply to 3.0?

Changes to 3.0 version:

*  Removed Call `AddMemoryCache` in `ConfigureServices`: - That's not required. Anyone who needs to call `AddMemoryCache` in `ConfigureServices` will know how to do that.
* Added:
   If SizeLimit is not set, the cache grows without bound. The ASP.NET Core runtime does not trim the cache when system memory is low. Apps much be architected to:

    Limit cache growth.
    Call Compact or Remove when available memory is limited:
* Added section [MemoryCache.Compact](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#memorycachecompact)
* Ported all the code to 3.0 and tested. It all works fine. So the 3.0 version of this shows the 3.0 `StartUp`.

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@mkArtakMSFT  given the 3.0 changes are trivial can you sign off on this PR or suggest a reviewer?

Changes to 3.0 version:

*  Removed Call `AddMemoryCache` in `ConfigureServices`: - That's not required. Anyone who needs to call `AddMemoryCache` in `ConfigureServices` will know how to do that.
* Ported all the code to 3.0 and tested. It all works fine. So the 3.0 version of this shows the 3.0 `StartUp`.

**Comment by mkArtakMSFT** [comment]
@anurse can you assign somebody to review this?

**Comment by BrennanConroy** [comment]
Could you explain what parts are new and need reviewing?

It looks like you split the docs into >=3.0 and <3.0 and a lot of the content was copied. So what is actually new?

**Comment by Rick-Anderson** [comment]
@BrennanConroy  can you give me a couple days, I've decided to do some more work on it.

> So what is actually new?

See `Changes to 3.0 version:` at the top of the page.

**Comment by Rick-Anderson** [comment]
@BrennanConroy  ready for review. I don't  think you need to review much unless there have been changes (beyond bug fixes) to caching for 3.0.

**Comment by BrennanConroy** [comment]
>Removed Call AddMemoryCache in ConfigureServices: - That's not required. Anyone who needs to call AddMemoryCache in ConfigureServices will know how to do that.

I'm a bit confused by this. How will they know? We aren't telling them anymore.

File: aspnetcore/performance/caching/memory/3.0sample/RPcache/Services/MyMemoryCache.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿using Microsoft.Extensions.Caching.Memory;
+
+namespace RPcache.Services
+{
+    #region snippet
+    // using Microsoft.Extensions.Caching.Memory;
+    public class MyMemoryCache 
+    {
+        public MemoryCache Cache { get; set; }
```
Comment (BrennanConroy) [inline aspnetcore/performance/caching/memory/3.0sample/RPcache/Services/MyMemoryCache.cs]: Make the setter private

File: aspnetcore/performance/caching/memory.md [hunk 2]
```diff
@@ -9,25 +9,183 @@ uid: performance/caching/memory
...
+* When using a callback to repopulate a cache item:
+
+  * Multiple requests can find the cached key value empty because the callback hasn't completed.
+  * This can result in several threads repopulating the cached item.
+
+* When one cache entry is used to create another, the child copies the parent entry's expiration tokens and time-based expiration settings. The child isn't expired by manual removal or updating of the parent entry.
+
+* Use<xref:Microsoft.Extensions.Caching.Memory.ICacheEntry.PostEvictionCallbacks> to set the callbacks that will be fired after the cache entry is evicted from the cache.
```
Comment (BrennanConroy) [inline aspnetcore/performance/caching/memory.md]:
```suggestion
* Use <xref:Microsoft.Extensions.Caching.Memory.ICacheEntry.PostEvictionCallbacks> to set the callbacks that will be fired after the cache entry is evicted from the cache.
```

**Comment by Rick-Anderson** [comment]
@BrennanConroy  sorry I missed a couple issues tied to this doc.

* Select [MemoryCacheEntryOptions](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#memorycacheentryoptions) and review the two samples above, starting at *The following code gets or creates a cached item with absolute expiration:*
* Select [Additional notes](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#additional-notes) and review the first bullet, ie *Expiration doesn't happen in the background.*

If you like, I can add something about startup. Something like:

In-memory caching is a service that's referenced from your app using Dependency Injection. `Add the following:`  In `ConfigureServices`, the following API's call `AddMemoryCache` internally, so it's not necessary to call `AddMemoryCache`:

* `AddMvc`
* `AddMvcCore().AddRazorViewEngine`
* `AddResponseCaching`
* `AddControllersWithViews`
* `AddRazorPages`

**Comment by BrennanConroy** [comment]
> the following API's call `AddMemoryCache` internally

Ah, didn't realize this. I still feel like we should tell users to add it explicitly if they want to use `IMemoryCache`.

I don't think we want to get into the business of maintaining a list of which middleware includes other middleware automatically.

**Comment by Rick-Anderson** [comment]
>  I still feel like we should tell users to add it explicitly if they want to use `IMemoryCache`.

I'm not sure I understand.  Click on [Using IMemoryCache](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#using-imemorycache)
I'm using `IMemoryCache` without calling `AddMemoryCache`

**Comment by BrennanConroy** [comment]
>I'm using IMemoryCache without calling AddMemoryCache

Yeah, because you're using one of the middleware that magically adds it behind the scenes.

Imagine you're a developer who wants to use `IMemoryCache` but doesn't have an app with MVC or Razor Pages or any of those middleware. You are going to read the document and try to inject `IMemoryCache` into one of your services and it will fail. And then you'll file an issue saying the doc doesn't explain how to add `IMemoryCache` to DI.

Side-note: Why is the warning in 2.2 under `Using IMemoryCache` no longer there in 3.0? I'm assuming it still applies.

**Comment by Rick-Anderson** [comment]
> Imagine you're a developer who wants to use IMemoryCache but doesn't have an app with MVC or Razor Pages or any of those middleware.

I added the following to the **Additional Notes** section. I don't want to pollute the intro with this fairly clumsy information. Eilon often said, those .01% who want to do that will figure out how to do it. Not it's explicitly noted in the article. BTW, we had two issues created where customers complained they didn't need to call `AddMemoryCache`.

> Side-note: Why is the warning in 2.2 under `Using IMemoryCache` no longer there in 3.0?

Fixed. Thanks for catching that.

**Comment by BrennanConroy** [comment]
Ok, I'm happy with the little blurb :+1:

**Pull request merged**